### PR TITLE
Add a MRO walker for slot methods and Convert _csv to use PEP 573

### PR DESCRIPTION
**Pull request by encukou**

### CODE REVIEW

File: Modules/_csv.c [hunk 1]
```diff
@@ -32,16 +37,28 @@ get_csv_state(PyObject *module)
 static int
 _csv_clear(PyObject *m)
 {
-    Py_CLEAR(get_csv_state(m)->error_obj);
-    Py_CLEAR(get_csv_state(m)->dialects);
+    _csvstate *m_state = PyModule_GetState(m);
+    if (m_state) {
```
Comment (shihai1991) [inline Modules/_csv.c:41]: After #18738, we don't need to identify `state` 's value.

File: Modules/_csv.c [hunk 2]
```diff
@@ -32,16 +37,28 @@ get_csv_state(PyObject *module)
...
     return 0;
 }
 
 static int
 _csv_traverse(PyObject *m, visitproc visit, void *arg)
 {
-    Py_VISIT(get_csv_state(m)->error_obj);
-    Py_VISIT(get_csv_state(m)->dialects);
+    _csvstate *m_state = PyModule_GetState(m);
+    if (m_state) {
```
Comment (shihai1991) [inline Modules/_csv.c:55]: dittio.

File: Modules/_csv.c [hunk 3]
```diff
@@ -882,13 +884,15 @@ Reader_iternext(ReaderObj *self)
...
 {
+    PyTypeObject *tp = Py_TYPE(self);
     PyObject_GC_UnTrack(self);
     Py_XDECREF(self->dialect);
     Py_XDECREF(self->input_iter);
     Py_XDECREF(self->fields);
     if (self->field != NULL)
         PyMem_Free(self->field);
     PyObject_GC_Del(self);
+    Py_DECREF(tp);
```
Comment (shihai1991) [inline Modules/_csv.c:904]: `Py_Type()` can not increase the refcount, so we don't need to use `Py_DECREF(tp)` in here, no?

File: Objects/typeobject.c [hunk 4]
```diff
@@ -3232,6 +3232,31 @@ PyType_GetModuleState(PyTypeObject *type)
...
+            break;
+        }
+        PyHeapTypeObject *ht = (PyHeapTypeObject*)super;
+        if (ht->ht_module && PyModule_GetDef(ht->ht_module) == def) {
+            return ht->ht_module;
+        }
+    }
+    PyErr_Format(
+        PyExc_TypeError,
+        "_PyType_GetModuleByDef: No superclass of '%s' has the given module",
```
Comment (shihai1991) [inline Objects/typeobject.c:3254]: Hm. If  type have no super class, we should judge the return result and call `PyType_GetModule()` again?

**Comment by encukou** [comment]
Thanks, Shi Hai!
I've made the change and posted this as https://github.com/python/cpython/pull/22835

**Pull request closed without merging**